packages/contracts/contracts/L1/ValidatorManager.sol

```diff
@@ -0,0 +1,637 @@
+            commissionMaxChangeRate,
+            assets
+        );
+    }
+
+    /**
+     * @inheritdoc IValidatorManager
+     */
+    function startValidator() external {
+        if (getStatus(msg.sender) != ValidatorStatus.CAN_START) revert ImproperValidatorStatus();
+
+        _validatorTree.insert(msg.sender, uint120(ASSET_MANAGER.reflectiveWeight(msg.sender)));
+
+        emit ValidatorStarted(msg.sender, block.timestamp);
+    }
+
+    /**
+     * @inheritdoc IValidatorManager
+     */
+    function afterSubmitL2Output(uint256 outputIndex) external onlyL2OutputOracle {
+        _distributeReward();
+
+        address submitter = L2_ORACLE.getSubmitter(outputIndex);
+        if (submitter == _nextPriorityValidator) {
+            _resetNoSubmissionCount(submitter);
+        } else {
+            _tryJail();
+        }
+
+        // Select the next priority validator.
+        _updatePriorityValidator();
+    }
+
+    /**
+     * @inheritdoc IValidatorManager
+     */
+    function changeCommissionRate(uint8 newCommissionRate) external {
+        if (getStatus(msg.sender) <= ValidatorStatus.INACTIVE) revert ImproperValidatorStatus();
+
+        Validator storage validatorInfo = _validatorInfo[msg.sender];
+
+        if (
+            validatorInfo.commissionRateChangedAt + COMMISSION_RATE_MIN_CHANGE_SECONDS >
+            block.timestamp
+        ) revert NotElapsedCommissionChangePeriod();
+
+        if (newCommissionRate > COMMISSION_RATE_DENOM) revert MaxCommissionRateExceeded();
+
+        uint8 oldCommissionRate = validatorInfo.commissionRate;
+        if (newCommissionRate == oldCommissionRate) revert SameCommissionRate();
+
+        uint8 changeRange = newCommissionRate > oldCommissionRate
+            ? newCommissionRate - oldCommissionRate
+            : oldCommissionRate - newCommissionRate;
+        if (changeRange > validatorInfo.commissionMaxChangeRate)
+            revert CommissionChangeRateExceeded();
+
+        validatorInfo.commissionRate = newCommissionRate;
+        validatorInfo.commissionRateChangedAt = uint128(block.timestamp);
+
+        emit ValidatorCommissionRateChanged(msg.sender, oldCommissionRate, newCommissionRate);
+    }
+
+    /**
+     * @inheritdoc IValidatorManager
+     */
+    function tryUnjail() external {
+        if (getStatus(msg.sender) != ValidatorStatus.IN_JAIL) revert ImproperValidatorStatus();
+
+        if (_jail[msg.sender] > block.timestamp) revert NotElapsedJailPeriod();
+
+        delete _jail[msg.sender];
+
+        _resetNoSubmissionCount(msg.sender);
+
+        emit ValidatorUnjailed(msg.sender);
+    }
+
+    /**
+     * @notice Slash KRO at the vault of the validator.
+     *
+     * @param loser       Address of the loser at the challenge.
+     * @param winner      Address of the winner at the challenge.
+     * @param outputIndex The index of output challenged.
+     */
+    function slash(address loser, address winner, uint256 outputIndex) external onlyColosseum {
```

Comment:
We don't need to jail the loser?

Reply:
I think it's good idea, but following PR(Colosseum PR) quite changes `slash` function, so I will do it in the following PR.